apps/consulting/posts/performance-for-image-processing-with-cucim.md

```diff
@@ -0,0 +1,274 @@
+the [Open Microscopy Environment (OME)][ome] project).
+
+## `cucim.skimage` Architecture
+
+The `cucim.skimage` module is currently implemented using [CuPy][cupy]. CuPy
+provides a GPU-based implementation of the NumPy API as well as a useful subset
+of the SciPy API (FFTs, sparse matrices and scipy.ndimage functions). As part of
+this work we also contributed many image processing functions from the SciPy API
+back upstream to CuPy itself (see, for example, the image affine transform
+benchmarks in the [complementary NVIDIA post][nvidia companion post]). These
+capabilities have made it possible to implement a substantial fraction of the
+`scikit-image` API in a relatively short time. At this point more than half of
+the `scikit-image` API has been implemented and we hope to continue to expand
+our API coverage over time.
+
+## Adding GPU Support to Existing CPU-Based Code
+
+Given that the `cucim.skimage` module replicates the `scikit-image` API, it is
+possible to run many existing [`scikit-image` examples][skimage examples] with
+minimal modification. To demonstrate this, we provide
+[example IPython notebooks][example notebooks] that have been adapted from
+gallery examples. A concrete example involves
+[Gabor filtering on both the CPU and GPU][gabor example]. In this example we
+have intentionally rerun the same example on both the CPU and GPU, with the only
+difference between the two cases being the following segment of code:
+
+```python
+if use_gpu:
+    from cupyx.scipy import ndimage as ndi
+    from cucim.skimage.util import img_as_float32
+    from cucim.skimage.filters import gabor_kernel
+
+    xp = cp
+    asnumpy = cp.asnumpy
+    device_name = "gpu"
+else:
+    from scipy import ndimage as ndi
+    from skimage.util import img_as_float32
+    from skimage.filters import gabor_kernel
+
+    xp = np
+    asnumpy = np.asarray
+    device_name = "cpu"
+
+```
+
+Here it can be seen that the primary difference is in the import statements and
+the use of a variable `xp` to represent either the CuPy or NumPy array module.
+The `asnumpy` function is defined to transfer data back from the GPU for
+visualization with Matplotlib. Users inspecting the output on the CPU and GPU
+will see that it is equivalent and the ratio of the time spent on the CPU vs.
+GPU will be displayed.
+
+Please note that because CuPy does just-in-time compilation of CUDA kernels the
+first time they are called, it is expected that the very first time a given
+example is run, it will have compilation overhead. On repeated executions,
+previously compiled kernels will be fetched either from an in-memory cache (or
+from CuPy’s on disk kernel cache in the case of a new Python session).
+
+The necessary changes typically involve:
+
+1. Use `cupy.asarray` to transfer array inputs to cuCIM functions from the host
+   to the GPU.
+2. Replace functions imported from CPU-based library with GPU equivalents as in
+   the following table:
+
+| CPU Module | GPU Module      |
```

Comment:
@noatamir / @gabalafou / @trallard -- do any of you know how we could constrain the max width of this table (say, to 50% of the story column width) via some sort of MDX syntax?

Reply:
Never mind, adding the table as raw, Tailwind-styled HTML worked well, once I added `.md` and `.mdx` to the Tailwind content search config.